dom/src/utils.ts

```diff
@@ -46,18 +46,51 @@ export function getValidNode(
   return domElement;
 }
 
-/**
- * The full scope of a namespace is the "absolute path" of scopes from
- * parent to child. This is extracted from the namespace, filter only for
- * scopes in the namespace.
- */
-export function getFullScope(namespace: Array<String>): string {
+export function getSelectors(namespace: Array<Scope>): string {
   return namespace
-    .filter(c => c.indexOf(SCOPE_PREFIX) > -1)
-    .map(c => c.replace(SCOPE_PREFIX, ''))
-    .join('-');
+    .filter(n => n.type === 'selector')
+    .map(n => n.scope)
+    .join(' ');
+}
+
+export function isEqualNamespace(
+  a: Array<Scope> | undefined,
+  b: Array<Scope> | undefined,
+): boolean {
+  if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) {
+    return false;
+  }
+  for (let i = 0; i < a.length; i++) {
+    if (a[i].type !== b[i].type || a[i].scope !== b[i].scope) {
+      return false;
+    }
+  }
+  return true;
+}
+
+export function getTotalIsolatedScope(namespace: Array<Scope>): Array<Scope> {
```

Comment:
This function is in a hot path: EventDelegator.onEvent -> EventDelegator.buildPath -> getTotalIsolatedScope, so we should avoid slice+splice which are not the most efficient way of writing this function. We can build the result array by construction, instead of cloning then removing.